datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java

```diff
@@ -728,6 +731,73 @@ private void removeTaskNode(String instance, String name) {
     _zkclient.deleteRecursive(instancePath);
   }
 
+  /**
+   * Update the task assignment for the given instances and issue assignment tokens for stopping tasks.
+   * Assignment tokens are used by the follower hosts to signal the leader that they finished handling their assignments.
+   *
+   * In addition to the work done by updateAllAssignments, this method writes to
+   * <ol>
+   *    *  <li>{@code /<cluster>/dms/<datastream-name>/assignmentTokens/<token1>,<token2>...}</li>
+   * </ol>
+   */
+  public void updateAllAssignmentsAndIssueTokens(Map<String, List<DatastreamTask>> assignmentsByInstance,
+      List<DatastreamGroup> stoppingDatastreamGroups) {
+    Validate.notNull(assignmentsByInstance);
+    Validate.notNull(stoppingDatastreamGroups);
+
+    Map<String, Set<DatastreamTask>> currentAssignment = getAllAssignedDatastreamTasks();
```

Comment:
The java doc of `getAllAssignedDatastreamTasks` says that there may be uncleaned dead instance – task assignments left which will be cleaned up later in the `handleLeaderDoAssignment` routine (line #1245).

So, if the leader issues tokens to those dead hosts, those would never be picked up?

Reply:
Even if they never get picked up, leader will end up revoking them. I'll see if it's possible to distinguish these dead instances and exclude them from the error log message and metrics (no need to alert for dead instances). Those changes are in the next PR.